pkg/tsdb/cloudwatch/get_metric_query_batches.go

```diff
@@ -13,72 +13,79 @@ var nonWordRegex = regexp.MustCompile(`\W+`)
 // getMetricQueryBatches separates queries into batches if necessary. Metric Insight queries cannot run together, and math expressions must be run
 // with all the queries they reference.
 func getMetricQueryBatches(queries []*models.CloudWatchQuery, logger log.Logger) [][]*models.CloudWatchQuery {
-	metricInsightIndices := []int{}
-	mathIndices := []int{}
-	for i, query := range queries {
-		switch query.GetGetMetricDataAPIMode() {
-		case models.GMDApiModeSQLExpression:
-			metricInsightIndices = append(metricInsightIndices, i)
-		case models.GMDApiModeMathExpression:
-			mathIndices = append(mathIndices, i)
-		default:
-		}
-	}
-	// We only need multiple batches if there are multiple metrics insight queries
-	if len(metricInsightIndices) <= 1 {
+	if !hasMultipleMetricInsights(queries) {
 		return [][]*models.CloudWatchQuery{queries}
 	}
 
 	logger.Debug("Separating queries into batches")
-	// Map ids to their queries
-	idToIndex := map[string]int{}
-	for i, query := range queries {
-		if query.Id != "" {
-			idToIndex[query.Id] = i
+
+	// set up list of math expression queries since below we loop over them to get what query IDs they reference
+	var mathQueries []*models.CloudWatchQuery
+	for _, query := range queries {
+		if query.GetGetMetricDataAPIMode() == models.GMDApiModeMathExpression {
+			mathQueries = append(mathQueries, query)
 		}
 	}
 
-	// Find and track which queries are referenced by math queries
-	queryReferences := make([][]int, len(queries))
-	isReferenced := make([]bool, len(queries))
-	for _, idx := range mathIndices {
-		tokens := nonWordRegex.Split(queries[idx].Expression, -1)
-		references := []int{}
-		for _, token := range tokens {
-			ref, found := idToIndex[token]
+	// put queries into a set in order to facilitate lookup below
+	idToQuery := make(map[string]*models.CloudWatchQuery, len(queries))
+	for _, q := range queries {
+		idToQuery[q.Id] = q
+	}
+
+	// gets query IDs which are referenced in math expressions
+	mathQueryIdToReferences := make(map[string][]*models.CloudWatchQuery)
```

Comment:
A tentative suggestion: `mathQueryIdToReferences` -> `mathQueryIdsWithReferences `

Reply:
The naming of this seems to denote that this structure is used as a lookup/set later on line 70.  I am not strongly opposed to your suggestion! On the other hand, I I think this matches the same naming as `idToQuery`. I think I'm inclined to keep it as-is, but totally open to discussing this further or trying other ideas!